src/metabase/db/custom_migrations.clj

```diff
@@ -352,3 +361,88 @@
+  ;; There will inevitably be extra entries in visualization_settings.column_settings that don't match field refs in result_metadata, but that's ok.
+  (let [add-join-aliases
+        (fn [card]
+          (let [join-aliases (->> (get-in card ["dataset_query" "query" "joins"])
+                                  (map #(get % "alias"))
+                                  set)]
+            (if (seq join-aliases)
+              (update (get card "visualization_settings") "column_settings"
+                      (fn [column_settings]
+                        (let [copies-with-join-alias (into {}
+                                                           (mapcat (fn [[k v]]
+                                                                     (match (vec (json/parse-string k))
+                                                                       ["ref" ["field" id opts]]
+                                                                       (for [alias join-aliases]
+                                                                         [(json/generate-string ["ref" ["field" id (assoc opts "join-alias" alias)]]) v])
+                                                                       _ '()))
+                                                                   column_settings))]
+                          ;; existing column settings should take precedence over the copies in case there is a conflict
+                          (merge copies-with-join-alias column_settings))))
+              card)))
+        update-one!
+        (fn [revision]
+          (let [card (json/parse-string (:object revision))]
+            (when (not= (get card "query_type") "native") ; native queries won't have join aliases, so we can exclude them straight away
+              (let [updated (add-join-aliases card)]
+                (when (not= updated (get "visualization_settings" card))
+                  (t2/query {:update :revision
+                             :set {:object (json/generate-string (assoc card "visualization_settings" updated))}
+                             :where [:= :id (:id revision)]}))))))]
+    (run! update-one! (t2/reducible-query {:select [:*]
+                                        :from   [:revision]
+                                        :where  [:and
+                                                 ;; only include cards with field refs in column_settings
+                                                 [:or
+                                                  [:like :object "%ref\\\\\",[\\\\\"field%"]
+                                                  [:like :object "%ref\\\\\\\",[\\\\\\\"field%"]]
+                                                 ;; only include cards with joins
+                                                 [:like :object "%joins%"]
+                                                 [:= :model "Card"]]})))
+  ;; Reverse migration
+  (let [update-one!
```

Comment:
I might miss the context behind this from the previous PRs, but why do we need to reverse it?

Reply:
Good q. You could change my mind on whether we should include this or not. We actually don't _have_ to reverse it, since there will always be one entry in column_settings with no join-alias that is identical to the before the migration. But I thought we should avoid having extra entries if we can. It's unlikely to cause any issues either way though.